### PR TITLE
Adding hero example

### LINKED ISSUE #69477: Adopt a Widget - Hero
The [Hero](https://master-api.flutter.dev/flutter/widgets/Hero-class.html) widget needs:

- A code / DartPad sample

For details about how to participate, see [flutter.dev/adoptawidget](https://flutter.dev/adoptawidget)

### PR DESCRIPTION
**Pull request by Abhishek01039**
AdoptAWidget: Hero

This pull request is:

 - [x] A code sample
 - [x] More references
 - [x] More explanation
closes https://github.com/flutter/flutter/issues/69477

### CODE REVIEW

**Comment by Abhishek01039** [comment]
Hello @johnpryan 
can you please review this PR?
Thanks

**Review by Piinks — COMMENTED** [review]
Hi @Abhishek01039 it looks like a few tests are failing here, can you take a look?

**Comment by Abhishek01039** [comment]
Hello @Piinks I just added the example.
Then also I need to add test ?

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 1]
```diff
@@ -101,6 +101,104 @@ enum HeroFlightDirection {
 ///
 /// {@youtube 560 315 https://www.youtube.com/watch?v=Be9UH1kXFDw}
 ///
+/// {@tool dartpad --template=stateless_widget_material}
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/heroes.dart:104]:
```suggestion
/// {@tool dartpad --template=stateful_widget_scaffold_center}
```

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 2]
```diff
@@ -101,6 +101,104 @@ enum HeroFlightDirection {
...
+///    return HeroAnimationRecipe();
+///  }
+///
+///  class HeroAnimationRecipe extends StatefulWidget {
+///
+///    @override
+///    _HeroAnimationRecipeState createState() => _HeroAnimationRecipeState();
+///  }
+///
+///  class _HeroAnimationRecipeState extends State<HeroAnimationRecipe> {
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/heroes.dart:124]: I think all of this can be removed when you use the right template.
```suggestion
```

Comment (Abhishek01039) [reply packages/flutter/lib/src/widgets/heroes.dart:124]: But we need two widgets for navigation

File: packages/flutter/lib/src/widgets/heroes.dart (hunk 1 above)
Comment (Abhishek01039) [reply packages/flutter/lib/src/widgets/heroes.dart:104]: This only covers one widget

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 3]
```diff
@@ -101,6 +101,107 @@ enum HeroFlightDirection {
...
+///      );
+///    }
+///
+///    Widget _blueDetailRectangle() {
+///     return Container(
+///      width: 200,
+///       height: 200,
+///       color: Colors.blue,
+///     );
+///    }
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/heroes.dart:142]: Could these methods (rectangle, detail) be reduced to just one that takes a size parameter?

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 4]
```diff
@@ -101,6 +101,107 @@ enum HeroFlightDirection {
...
+/// ```
+///
+/// ```dart preamble
+///  class HeroAnimationRecipe extends StatefulWidget {
+///
+///    @override
+///    _HeroAnimationRecipeState createState() => _HeroAnimationRecipeState();
+///  }
+///
+///  class _HeroAnimationRecipeState extends State<HeroAnimationRecipe> {
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/heroes.dart:126]: I think all of this can be removed? The outer class where you have just the other class in the build method does not seem necessary.
```suggestion
```

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 5]
```diff
@@ -101,6 +101,107 @@ enum HeroFlightDirection {
...
+///             title: Text('Tap on the icon to view hero animation transition.'),
+///            ),
+///          ],
+///        ),
+///      );
+///    }
+///
+///    void _gotoDetailsPage(BuildContext context) {
+///      Navigator.of(context).push(MaterialPageRoute(
+///        builder: (ctx) => Scaffold(
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/heroes.dart:179]:
```suggestion
///        builder: (BuildContext context) => Scaffold(
```

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 6]
```diff
@@ -101,6 +101,107 @@ enum HeroFlightDirection {
...
+///          body: Center(
+///            child: Column(
+///               mainAxisAlignment: MainAxisAlignment.center,
+///               children: <Widget>[
+///                Hero(
+///                  tag: 'hero-rectangle',
+///                  child: _blueDetailRectangle(),
+///                ),
+///                Text(
+///                   'This is where you can see details about the list item tapped at previous page.',
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/heroes.dart:192]: There are no 'details', so this message is a little confusing.

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 7]
```diff
@@ -101,6 +101,107 @@ enum HeroFlightDirection {
...
+///
+///    void _gotoDetailsPage(BuildContext context) {
+///      Navigator.of(context).push(MaterialPageRoute(
+///        builder: (ctx) => Scaffold(
+///          appBar: AppBar(
+///            title: Text('second Page'),
+///          ),
+///          body: Center(
+///            child: Column(
+///               mainAxisAlignment: MainAxisAlignment.center,
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/heroes.dart:185]: I think the indentation is off in a few places. Be sure to check the style guide, flutter does not use dartfmt.

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 8]
```diff
@@ -101,6 +101,107 @@ enum HeroFlightDirection {
...
 /// {@youtube 560 315 https://www.youtube.com/watch?v=Be9UH1kXFDw}
 ///
+/// {@tool dartpad --template=stateful_widget_scaffold_center}
+/// This sample shows a `Hero` with List item along with Icon: here we will be using a
+/// rectangle widget for icon. Icon widget in list item at first page , with Hero
+/// widget in list and Enlarged icon to display on Second Page.
+///
+/// Navigator.of(context).push() is needed to navigate to second page from
+/// first page using the SAME tag Making sure both widgets have same tag. Hero
+/// transitions use the same tag to identify the start and destination widgets.
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/heroes.dart:111]:
```suggestion
/// This sample shows a [Hero] used within a [ListTile]. Tapping on the Hero-wrapped rectangle triggers a hero
/// animation as a new [MaterialPageRoute] is pushed. Both the size and location of the rectangle animates.
///
/// Both widgets use the same [Hero.tag]. The Hero widget uses the matching tags to identify and execute this
/// animation.
```

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 9]
```diff
@@ -101,6 +101,107 @@ enum HeroFlightDirection {
...
+///     return Scaffold(
+///       appBar: AppBar(
+///         title: Text('Hero Animation'),
+///       ),
+///       body: buildDemoWidget(context),
+///     );
+///    }
+///
+///    Widget buildDemoWidget(BuildContext context) {
+///      return Center(
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/heroes.dart:155]: The template will have this wrapped in a center already.

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 10]
```diff
@@ -101,6 +101,107 @@ enum HeroFlightDirection {
...
+///      width: 200,
+///       height: 200,
+///       color: Colors.blue,
+///     );
+///    }
+///
+///    @override
+///    Widget build(BuildContext context) {
+///     return Scaffold(
+///       appBar: AppBar(
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/heroes.dart:147]: The template already creates a Scaffold and an AppBar.

**Review by Piinks — CHANGES_REQUESTED** [review]

File: packages/flutter/lib/src/widgets/heroes.dart (hunk 5 above)
Comment (Abhishek01039) [reply packages/flutter/lib/src/widgets/heroes.dart:179]: Ah, Thanks @Piinks

File: packages/flutter/lib/src/widgets/heroes.dart (hunk 8 above)
Comment (Abhishek01039) [reply packages/flutter/lib/src/widgets/heroes.dart:111]: Thanks @Piinks

File: packages/flutter/lib/src/widgets/heroes.dart (hunk 6 above)
Comment (Abhishek01039) [reply packages/flutter/lib/src/widgets/heroes.dart:192]: I have removed this Text widget.

File: packages/flutter/lib/src/widgets/heroes.dart (hunk 3 above)
Comment (Abhishek01039) [reply packages/flutter/lib/src/widgets/heroes.dart]: Thanks @Piinks

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 11]
```diff
@@ -101,6 +101,107 @@ enum HeroFlightDirection {
 ///
 /// {@youtube 560 315 https://www.youtube.com/watch?v=Be9UH1kXFDw}
 ///
+/// {@tool dartpad --template=stateful_widget_scaffold_center}
+/// This sample shows a `Hero` with List item along with Icon: here we will be using a
+/// rectangle widget for icon. Icon widget in list item at first page , with Hero
+/// widget in list and Enlarged icon to display on Second Page.
```
Comment (johnpryan) [inline packages/flutter/lib/src/widgets/heroes.dart]: This is a little confusing, `Icon` typically means the `Icon` widget, but in this sample you are using a Container with a color.

```suggestion
/// {@tool dartpad --template=stateful_widget_scaffold_center}
/// This sample displays shows a `Hero` that flies between the two screens
/// and grows to fit the layout on the second page.
```

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 12]
```diff
@@ -101,6 +101,107 @@ enum HeroFlightDirection {
...
+/// ```dart preamble
+///  class HeroAnimationRecipe extends StatefulWidget {
+///
+///    @override
+///    _HeroAnimationRecipeState createState() => _HeroAnimationRecipeState();
+///  }
+///
+///  class _HeroAnimationRecipeState extends State<HeroAnimationRecipe> {
+///
+///    Widget _blueIconRectangle() {
```
Comment (johnpryan) [inline packages/flutter/lib/src/widgets/heroes.dart]: Rename this to `_blueListRectangle`, since it's not an icon in this example.

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 13]
```diff
@@ -101,6 +101,107 @@ enum HeroFlightDirection {
...
+///    Widget buildDemoWidget(BuildContext context) {
+///      return Center(
+///        child: Column(
+///          crossAxisAlignment: CrossAxisAlignment.start,
+///          children: <Widget>[
+///            SizedBox(
+///              height: 20.0,
+///            ),
+///            ListTile(
+///             leading: GestureDetector(
```
Comment (johnpryan) [inline packages/flutter/lib/src/widgets/heroes.dart:152]: `ListTile` already has an `onTap` callback so this can be removed.

**Review by johnpryan — CHANGES_REQUESTED** [review]

File: packages/flutter/lib/src/widgets/heroes.dart (hunk 13 above)
Comment (Abhishek01039) [reply packages/flutter/lib/src/widgets/heroes.dart]: Ah! Thanks

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 14]
```diff
@@ -101,6 +100,71 @@ enum HeroFlightDirection {
 ///
 /// {@youtube 560 315 https://www.youtube.com/watch?v=Be9UH1kXFDw}
 ///
+/// {@tool dartpad --template=stateful_widget_scaffold_center}
+/// This sample shows a [Hero] used within a [ListTile]. Tapping on the Hero-wrapped rectangle triggers a hero
+/// animation as a new [MaterialPageRoute] is pushed. Both the size and location of the rectangle animates.
+///
+/// Both widgets use the same [Hero.tag]. The Hero widget uses the matching tags to identify and execute this
+/// animation.
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/heroes.dart:108]: nit: place keep the lines here to approx 60 char.

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 15]
```diff
@@ -101,6 +100,71 @@ enum HeroFlightDirection {
...
+///
+/// ```dart
+///  Widget build(BuildContext context) {
+///    return buildDemoWidget(context);
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/heroes.dart:112]: This indirection via the buildDemoWidget is not necessary. Just put the content of buildDemoWidget right into the build method.

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 16]
```diff
@@ -101,6 +100,71 @@ enum HeroFlightDirection {
...
+///  }
+/// ```
+///
+/// ```dart preamble
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/heroes.dart:116]: remove these three lines. All the code should be in ```dart

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 17]
```diff
@@ -101,6 +100,71 @@ enum HeroFlightDirection {
...
+///          height: 20.0,
+///        ),
+///        ListTile(
+///          leading: Hero(
+///              tag: 'hero-rectangle',
+///              child: _blueRectangle(Size(50,50)),
+///            ),
+///            onTap: () => _gotoDetailsPage(context),
+///          ),
+///          title: Text('Tap on the icon to view hero animation transition.'),
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/heroes.dart:139]: formatting and parentheses:

```suggestion
///        ListTile(
///          leading: Hero(
///            tag: 'hero-rectangle',
///            child: _blueRectangle(Size(50,50)),
///          ),
///          onTap: () => _gotoDetailsPage(context),
///          title: Text('Tap on the icon to view hero animation transition.'),
///        ),
```

File: packages/flutter/lib/src/widgets/heroes.dart [hunk 18]
```diff
@@ -152,10 +216,10 @@ class Hero extends StatefulWidget {
     this.placeholderBuilder,
     this.transitionOnUserGestures = false,
     required this.child,
-  }) : assert(tag != null),
-       assert(transitionOnUserGestures != null),
-       assert(child != null),
-       super(key: key);
+  })   : assert(tag != null),
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/heroes.dart:219]: revert all these unrelated formatting changes! (here and below)

File: packages/flutter/lib/src/widgets/heroes.dart (hunk 15 above)
Comment (Abhishek01039) [reply packages/flutter/lib/src/widgets/heroes.dart]: Done. Thanks @goderbauer

File: packages/flutter/lib/src/widgets/heroes.dart (hunk 16 above)
Comment (Abhishek01039) [reply packages/flutter/lib/src/widgets/heroes.dart]: Done. Thanks

File: packages/flutter/lib/src/widgets/heroes.dart (hunk 17 above)
Comment (Abhishek01039) [reply packages/flutter/lib/src/widgets/heroes.dart]: Done Thanks

File: packages/flutter/lib/src/widgets/heroes.dart (hunk 18 above)
Comment (Abhishek01039) [reply packages/flutter/lib/src/widgets/heroes.dart]: Done Thanks

**Review by goderbauer — APPROVED** [review]
LGTM

**Review by johnpryan — APPROVED** [review]

**Comment by Piinks** [comment]
The google testing failure here is unrelated, setting to green.

**Review by Piinks — APPROVED** [review]

**Comment by Abhishek01039** [comment]
Hey, @Piinks 
can you please restart these tests?
so it might be passed !!

**Comment by shihaohong** [comment]
Actually, it's weird that the check is showing up. It turns out that it's already deleted (see https://github.com/flutter/flutter/issues/74462). The next time it gets stuck and everything else is green, let's just merge this PR.

**Comment by shihaohong** [comment]
Actually, it's weird that the check is showing up. It turns out that it's already deleted (see #74462). The next time it gets stuck and everything else is green, let's just merge this PR.

**Review by Piinks — COMMENTED** [review]
I think since that check does not actually exist now, this PR will need to be updated with the latest from master to resolve it.

**Comment by johnpryan** [comment]
@Abhishek01039 Thanks for participating in AdoptAWidget. We would like to send you a special thanks. If you are interested, please send an email to nyener at google dot com and provide a link to this issue.

**Comment by Abhishek01039** [comment]
Hello, @johnpryan 
can you please add this pull request here? https://flutter.dev/adoptawidget
Thanks
